### PR TITLE
Add example / doc / func tests for Torch support

### PR DESCRIPTION
**Pull request by goodsong81**
[Changes]
- [x] Add example script for Torch->ONNX XAI support
  - How to run
  ```console
  python examples/run_torch_onnx.py
  ```
- [x] Update documentations
  image
- [x] Add functional tests

[Future work]
- [ ] Optional torch installation

### CODE REVIEW

File: docs/source/user-guide.md
Comment (negvet) [inline docs/source/user-guide.md:1]: Please make sure that arguments have proper names, e.g. `target_explain_labels` -> `targets`

File: docs/source/user-guide.md [hunk 1]
```diff
@@ -316,21 +316,44 @@ As mentioned above, saliency map generation requires model inference. In the abo
...
-)  # type: ov.Model
+)
 
-# ***** Downstream task: user's code that infers model_xai and picks 'saliency_map' output *****
+# For PyTorch models, XAI head is inserted using the module hook mechanism internally
+# so that users could get additional saliency map without major changes in original inference pipeline.
+model: torch.nn.Module
+
+# Insert XAI head
+xai_model: torch.nn.Module = insert_xai(model=model, task=xai.Task.CLASSIFICATION)
```
Comment (negvet) [inline docs/source/user-guide.md]: please align this name with all other names of model + xai objects, it seems that in most cases it is called `model_xai`. not a must, but good to have

File: examples/run_torch_onnx.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    outputs = ov_model(image_norm)
+    logits = outputs["prediction"]  # BxC
+    saliency_maps = outputs["saliency_map"]  # BxCxhxw
+    probs = softmax(logits)
+    label = probs.argmax(axis=-1)[0]
+    logger.info(f"OpenVINO XAI model prediction: classes ({probs.shape[-1]}) -> label ({label}) -> prob ({probs[0, label]})")
+
+    # OpenVINO XAI model saliency map
+    saliency_maps = saliency_maps.squeeze(0)  # Cxhxw
+    mask = saliency_maps[label]  # hxw mask for the label
```
Comment (negvet) [inline examples/run_torch_onnx.py]: why `mask`?

File: README.md [hunk 3]
```diff
@@ -24,9 +24,9 @@
 ![OpenVINO XAI Concept](docs/source/_static/ovxai-concept.svg)
 
 **OpenVINO™ Explainable AI (XAI) Toolkit** provides a suite of XAI algorithms for visual explanation of
-[**OpenVINO™**](https://github.com/openvinotoolkit/openvino) Intermediate Representation (IR) models.
+[**OpenVINO™**](https://github.com/openvinotoolkit/openvino) as well as [**PyTorch**](https://pytorch.org) and [**ONNX**](https://onnx.ai/) models.
 
-Given **OpenVINO** models and input images, **OpenVINO XAI** generates **saliency maps**
+Given AI models and input images, **OpenVINO XAI** generates **saliency maps**
```
Comment (GalyaZalesskaya) [inline README.md:29]: @negvet We use the exact same sentence in all of the notebooks and all of their READMEs, so we shouldn't forget to update it as well

File: openvino_xai/api/api.py [hunk 4]
```diff
@@ -50,7 +50,6 @@ def insert_xai(
         explain_method=explain_method,
         target_layer=target_layer,
         embed_scaling=embed_scaling,
-        prepare_model=False,
```
Comment (negvet) [inline openvino_xai/api/api.py:53]: Please elaborate, why `prepare_model=False`?

File: README.md [hunk 5]
```diff
@@ -196,6 +193,28 @@ Original image | Explained image
 
 We can see that model is focusing on the body or skin area of the animals to tell if this image contains actual cheetahs.
 
+### Insert XAI head to your models
+
+Using the `insert_xai` API, we can insert XAI head to existing OpenVINO or PyTorch models directly and get additional saliency map output in the same inference pipeline.
```
Comment (GalyaZalesskaya) [inline README.md]:
```suggestion
Using the `insert_xai` API, we can insert XAI head to existing OpenVINO or PyTorch models directly and get additional "saliency_map" output in the same inference pipeline.
```

File: docs/source/user-guide.md [hunk 6]
```diff
@@ -316,21 +316,44 @@ As mentioned above, saliency map generation requires model inference. In the abo
...
 import openvino_xai as xai
+from openvino_xai.common.utils import softmax
+from openvino_xai.explainer.visualizer colormap, overlay
 
 
 # Create an ov.Model
-model = ov.Core().read_model("path/to/model.xml")  # type: ov.Model
+model: ov.Model = ov.Core().read_model("path/to/model.xml")
 
 # Insert XAI branch into the model graph
```
Comment (GalyaZalesskaya) [inline docs/source/user-guide.md]:
```suggestion
# Insert XAI branch into the OpenVINO model graph (IR)
```

File: docs/source/user-guide.md [hunk 7]
```diff
@@ -316,21 +316,44 @@ As mentioned above, saliency map generation requires model inference. In the abo
...
+    # target_layer="last_conv_node_name",  # target_layer - the node after which the XAI branch will be inserted, usually the last convolutional layer in the backbone. Defaults to None, by which the target layer is automatically detected
     embed_scaling=True,  # True by default. If set to True, the saliency map scale (0 ~ 255) operation is embedded in the model
     explain_method=xai.Method.RECIPROCAM,  # ReciproCAM is the default XAI method for CNNs
-)  # type: ov.Model
+)
 
-# ***** Downstream task: user's code that infers model_xai and picks 'saliency_map' output *****
+# For PyTorch models, XAI head is inserted using the module hook mechanism internally
+# so that users could get additional saliency map without major changes in original inference pipeline.
+model: torch.nn.Module
```
Comment (GalyaZalesskaya) [inline docs/source/user-guide.md]:
```suggestion
# Insert XAI branch into the Pytorch model
# XAI head is inserted using the module hook mechanism internally
# so that users could get additional saliency map without major changes in the original inference pipeline.
model: torch.nn.Module
```

Comment (GalyaZalesskaya) [reply docs/source/user-guide.md]: Can we somehow shorten the comment 335-336?

File: docs/source/user-guide.md [hunk 8]
```diff
@@ -316,21 +316,44 @@ As mentioned above, saliency map generation requires model inference. In the abo
...
+    saliency_maps = outputs["saliency_map"]  # BxCxhxw: additional per-class saliency map
+    probs = torch.softmax(logits, dim=-1)
+    label = probs.argmax(dim=-1)[0]
+
+# Torch XAI model saliency map
+saliency_maps = saliency_maps.numpy(force=True).squeeze(0)  # Cxhxw
+mask = saliency_maps[label]  # hxw mask for the label
+mask = colormap(mask[None, :])  # 1xhxw
+mask = cv2.resize(mask.squeeze(0), dsize=input_size)  # HxW
+result = overlay(mask, image)
```
Comment (GalyaZalesskaya) [inline docs/source/user-guide.md]: "mask" for me personally has the strong connection with instance segmentation. Can we use "map" or "sal_map" instead?

**Review by negvet — CHANGES_REQUESTED** [review]

File: examples/run_torch_onnx.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+def main(argv: list[str]):
+    parser = get_argument_parser()
+    args = parser.parse_args(argv)
+
+    # Load Torch model from timm
+    try:
+        model = timm.create_model(args.model_name, in_chans=3, pretrained=True)
+        logger.info(f"Model config: {model.default_cfg}")
+        logger.info(f"Model layers: {model}")
+    except Exception as e:
```
Comment (GalyaZalesskaya) [inline examples/run_torch_onnx.py:36]: Can you move this block and everything below in the separate function `insert_xai_to_torch_export_to_ir` and copy the insert part and create separate `insert_xai_to_torch_export_to onnx`. Please add doc strings as well. It will create similarity between our `run_` files.
If you think, it's not needed, then can you just add a docstring to main?

File: examples/run_torch_onnx.py [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+        label = probs.argmax(dim=-1)[0]
+    logger.info(f"Torch XAI model prediction: classes ({probs.shape[-1]}) -> label ({label}) -> prob ({probs[0, label]})")
+
+    # Torch XAI model saliency map
+    saliency_maps = saliency_maps.numpy(force=True).squeeze(0)  # Cxhxw
+    mask = saliency_maps[label]  # hxw mask for the label
+    mask = colormap(mask[None, :])  # 1xhxw
+    mask = cv2.resize(mask.squeeze(0), dsize=input_size)  # HxW
+    result = overlay(mask, image)
+    result = cv2.cvtColor(result, code=cv2.COLOR_RGB2BGR)
```
Comment (GalyaZalesskaya) [inline examples/run_torch_onnx.py]: "mask" -> "sal_map"
"result" -> "result_saliency_map"?

File: tests/func/test_torch_onnx_timm_full.py [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+    def test_insert_xai(self, model_id, fxt_output_root: Path):
+        # for skipped_model in NOT_SUPPORTED_BY_WB_MODELS.keys():
+        #    if skipped_model in model_id:
+        #        pytest.skip(reason=NOT_SUPPORTED_BY_WB_MODELS[skipped_model])
+
+        # for failed_model in SUPPORTED_BUT_FAILED_BY_WB_MODELS.keys():
+        #    if failed_model in model_id:
+        #        pytest.xfail(reason=SUPPORTED_BUT_FAILED_BY_WB_MODELS[failed_model])
+
+        # Load Torch model from timm
```
Comment (GalyaZalesskaya) [inline tests/func/test_torch_onnx_timm_full.py:102]: This comments can be removed probably

**Review by GalyaZalesskaya — COMMENTED** [review]
Great update! Thank you for your work!

File: openvino_xai/api/api.py (hunk 4 above)
Comment (goodsong81) [reply openvino_xai/api/api.py:53]: I removed this line, and then it's `prepare_model=True` by default. It's for auto detect `Method`.
The "CNNs->ViTs" fallback mechanism is implemented in the `WhiteBoxMethodFactory.create_method()` and it works only with `prepared_model=True` because we try to insert XAI head while model creation in `Method.__init__()` by calling `Method.prepare_model()`. Otherwise, users should always specify the `Method` for `insert_xai()` API.

File: tests/func/test_torch_onnx_timm_full.py (hunk 11 above)
Comment (goodsong81) [reply tests/func/test_torch_onnx_timm_full.py:102]: I'll keep them and refine or remove after the functional validation. :)

File: docs/source/user-guide.md
Comment (goodsong81) [reply docs/source/user-guide.md:1]: Done in https://github.com/openvinotoolkit/openvino_xai/pull/66/commits/44426944a60f027ab3978c9c5d4a45827e2c5aee

File: docs/source/user-guide.md (hunk 1 above)
Comment (goodsong81) [reply docs/source/user-guide.md]: Done in https://github.com/openvinotoolkit/openvino_xai/pull/66/commits/44426944a60f027ab3978c9c5d4a45827e2c5aee

File: examples/run_torch_onnx.py (hunk 2 above)
Comment (goodsong81) [reply examples/run_torch_onnx.py]: Refined in https://github.com/openvinotoolkit/openvino_xai/pull/66/commits/44426944a60f027ab3978c9c5d4a45827e2c5aee

File: docs/source/user-guide.md (hunk 8 above)
Comment (goodsong81) [reply docs/source/user-guide.md]: Done in https://github.com/openvinotoolkit/openvino_xai/pull/66/commits/44426944a60f027ab3978c9c5d4a45827e2c5aee

File: examples/run_torch_onnx.py (hunk 10 above)
Comment (goodsong81) [reply examples/run_torch_onnx.py]: Refined in https://github.com/openvinotoolkit/openvino_xai/pull/66/commits/44426944a60f027ab3978c9c5d4a45827e2c5aee

File: examples/run_torch_onnx.py (hunk 9 above)
Comment (goodsong81) [reply examples/run_torch_onnx.py:36]: Done in https://github.com/openvinotoolkit/openvino_xai/pull/66/commits/202b049f40609892d63637a8824730aaf47bc6ff

File: openvino_xai/api/api.py (hunk 4 above)
Comment (negvet) [reply openvino_xai/api/api.py:53]: But if you see the below line (number 57), there is one more prepare the model (without loading). 
In this case we do prepare model twice to be able to support "CNNs->ViTs" fallback mechanism?

If yes, for the future work, we might do some checks to avoid duplicate work?

**Review by GalyaZalesskaya — APPROVED** [review]
Looks great, thanks!

File: examples/run_torch_onnx.py (hunk 10 above)
Comment (GalyaZalesskaya) [reply examples/run_torch_onnx.py]: One small update: "saliency_image" -> "overlayed_saliency_map" (Can be done later)

Comment (goodsong81) [reply examples/run_torch_onnx.py]: Changed to `result_image` as we discussed.

File: openvino_xai/api/api.py (hunk 4 above)
Comment (goodsong81) [reply openvino_xai/api/api.py:53]: I tried to avoid duplicated insertion and compile for that reason.
image
Once XAI-inserted (self._model is not None) or compiled (self._model_compiled is not None),
prepare_model() would just return the previously processed model.

Comment (negvet) [reply openvino_xai/api/api.py:53]: Indeed, seems all good, thanks!

**Review by negvet — APPROVED** [review]
